internal/pkg/rpaas/k8s.go

```diff
@@ -2605,6 +2692,388 @@ func (m *k8sRpaasManager) DeleteUpstream(ctx context.Context, instanceName strin
+	}
+
+	// New traffic shaping validation rules:
+	// 1. Primary upstream cannot have any traffic shaping policy when it has canary binds
+	if len(args.CanaryBinds) > 0 && hasTrafficShapingPolicy(args.TrafficShapingPolicy) {
+		return &ValidationError{Msg: fmt.Sprintf("primary upstream '%s' cannot have traffic shaping policy when it has canary binds", args.PrimaryBind)}
+	}
+
+	// 2. If this is a canary bind, validate weight rule: only one canary per group can have weight > 0
+	if args.TrafficShapingPolicy.Weight > 0 && isReferencedAsCanary {
+		if err := m.validateCanaryWeightRule(instance.Spec.UpstreamOptions, args.PrimaryBind, args.TrafficShapingPolicy.Weight, "add"); err != nil {
+			return err
+		}
+	}
+
+	// Validate that header-value and header-pattern are mutually exclusive
+	if strings.TrimSpace(args.TrafficShapingPolicy.HeaderValue) != "" && strings.TrimSpace(args.TrafficShapingPolicy.HeaderPattern) != "" {
+		return &ValidationError{Msg: "header-value and header-pattern are mutually exclusive, please specify only one"}
+	}
+
+	// Validate that LoadBalance algorithm is one of the supported values
+	if args.LoadBalance != "" {
+		validAlgorithms := []v1alpha1.LoadBalanceAlgorithm{
+			v1alpha1.LoadBalanceRoundRobin,
+			v1alpha1.LoadBalanceConsistentHash,
+			v1alpha1.LoadBalanceEWMA,
+		}
+
+		isValid := false
+		for _, valid := range validAlgorithms {
+			if args.LoadBalance == valid {
+				isValid = true
+				break
+			}
+		}
+
+		if !isValid {
+			return &ValidationError{Msg: fmt.Sprintf("invalid loadBalance algorithm: %s. Valid values are: round_robin, chash, ewma", args.LoadBalance)}
+		}
+	}
+
+	// Validate that LoadBalanceHashKey is required when LoadBalance is "chash"
+	if args.LoadBalance == v1alpha1.LoadBalanceConsistentHash && strings.TrimSpace(args.LoadBalanceHashKey) == "" {
+		return &ValidationError{Msg: "loadBalanceHashKey is required when loadBalance is \"chash\""}
+	}
+
+	// Validate that LoadBalanceHashKey is not provided for non-chash algorithms
+	if args.LoadBalance != v1alpha1.LoadBalanceConsistentHash && strings.TrimSpace(args.LoadBalanceHashKey) != "" {
+		return &ValidationError{Msg: "loadBalanceHashKey is only valid when loadBalance is \"chash\""}
+	}
+
+	// Apply defaults to UpstreamOptions
+	applyUpstreamOptionsDefaults(&args)
+
+	upstreamOptions := v1alpha1.UpstreamOptions{
+		PrimaryBind:          args.PrimaryBind,
+		CanaryBinds:          args.CanaryBinds,
+		TrafficShapingPolicy: args.TrafficShapingPolicy,
+		LoadBalance:          args.LoadBalance,
+		LoadBalanceHashKey:   args.LoadBalanceHashKey,
+	}
+
+	instance.Spec.UpstreamOptions = append(instance.Spec.UpstreamOptions, upstreamOptions)
+	return m.patchInstance(ctx, originalInstance, instance)
+}
+
+func (m *k8sRpaasManager) UpdateUpstreamOptions(ctx context.Context, instanceName string, args UpstreamOptionsArgs) error {
```

Comment:
I think a lot of the validations between the `AddUpstreamOptions` and `UpdateUpstreamOptions` are equal, we could maybe extract them into a function as reuse it